opal/util/keyval_parse.c

```diff
@@ -276,33 +276,38 @@ static int save_param_name (void)
 
 static int add_to_env_str(char *var, char *val)
 {
-    int sz, varsz, valsz;
+    int sz, varsz, valsz, new_envsize;
     void *tmp;
 
     if (NULL == var) {
         return OPAL_ERR_BAD_PARAM;
     }
 
-    if (NULL != env_str) {
-        varsz = strlen(var);
-        valsz = (NULL != val) ? strlen(val) : 0;
-        sz = strlen(env_str)+varsz+valsz+2;
-        if (envsize <= sz) {
-            envsize *=2;
+    varsz = strlen(var);
+    valsz = (NULL != val) ? strlen(val) : 0;
+    sz = (NULL != env_str) ? strlen(env_str) : 0;
+    sz += varsz+valsz+2;
+    new_envsize = envsize;
+    while (new_envsize <= sz) {
+        new_envsize *=2;
+    }
 
-            tmp = realloc(env_str, envsize);
+    if (NULL != env_str) {
+        if (new_envsize > envsize) {
+            tmp = realloc(env_str, new_envsize);
             if (NULL == tmp) {
                 return OPAL_ERR_OUT_OF_RESOURCE;
             }
             env_str = tmp;
         }
         strcat(env_str, ";");
```

Comment:
Is there a `+1` missing somewhere to account for the `;` character?

Reply:
There's also a `strcat()` of an additional `=` character on old line 309.  Does that need to be counted, too?

Reply:
You're right, the original version didn't account for the terminating NUL byte. The chance of hitting that is pretty slim as allocations are made in multiples of 1K but I force-pushed an update  that makes accounting =, ;, and \0 explicit.